src/extension/IOController_Posix.hpp

```diff
@@ -35,237 +35,307 @@ namespace extension {
 
     class IOController : public Reactor {
     private:
+        /**
+         * @brief A task that is waiting for an IO event
+         */
         struct Task {
             Task() = default;
             // NOLINTNEXTLINE(google-runtime-int)
             Task(const fd_t& fd, short events, std::shared_ptr<threading::Reaction> reaction)
                 : fd(fd), events(events), reaction(std::move(reaction)) {}
 
+            /// @brief The file descriptor we are waiting on
             fd_t fd{-1};
+            /// @brief The events that are waiting to be fired
+            short waiting_events{0};  // NOLINT(google-runtime-int)
+            /// @brief The events that the task is interested in
             short events{0};  // NOLINT(google-runtime-int)
+            /// @brief The reaction that is waiting for this event
             std::shared_ptr<threading::Reaction> reaction{nullptr};
 
+            /**
+             * @brief Sorts the tasks by their file descriptor
+             *
+             * The tasks are sorted by file descriptor so that when we rebuild the list of file descriptors to poll we
+             * can assume that if the same file descriptor shows up multiple times it will be next to each other. This
+             * allows the events that are being watched to be or'ed together.
+             *
+             * @param other  the other task to compare to
+             *
+             * @return true  if this task is less than the other
+             * @return false if this task is greater than or equal to the other
+             */
             bool operator<(const Task& other) const {
                 return fd == other.fd ? events < other.events : fd < other.fd;
             }
         };
 
+        /**
+         * @brief Rebuilds the list of file descriptors to poll
+         *
+         * This function is called when the list of file descriptors to poll changes. It will rebuild the list of file
+         * descriptors used by poll
+         */
+        void rebuild_list() {
+            // Get the lock so we don't concurrently modify the list
+            const std::lock_guard<std::mutex> lock(tasks_mutex);
+
+            // Clear our fds to be rebuilt
+            fds.resize(0);
+
+            // Insert our notify fd
+            fds.push_back(pollfd{notify_recv, POLLIN, 0});
```

Comment:
This would create a new vector, this is intentionally done as a resize(0) followed by insertion again so that the std::vector object maintains it's backing arrays.